ethpm_cli/commands/auth.py

```diff
@@ -1,23 +1,104 @@
+import json
 from pathlib import Path
-import tempfile
 from typing import Any, Dict
 
 import eth_keyfile
 from eth_typing import Address
-from eth_utils import to_bytes
+from eth_utils import add_0x_prefix, to_bytes
 
+from ethpm_cli._utils.filesystem import atomic_replace
+from ethpm_cli._utils.input import parse_bool_flag
+from ethpm_cli._utils.logger import cli_logger
 from ethpm_cli._utils.xdg import get_xdg_ethpmcli_root
 from ethpm_cli.constants import KEYFILE_PATH
 from ethpm_cli.exceptions import AuthorizationError
 
+PRIVATE_KEY_WARNING = (
+    "Please be careful when using your private key, it is a sensitive piece of information.\n",
+    "~ ~ ~ Not your keys, not your crypto. ~ ~ ~\n",
+    "ethPM doesn't save your private key, but it is best practice to re-use an already encrypted ",
+    "keyfile and link it with `ethpm auth link` rather than regenerating a new encrypted keyfile.",
+)
+
+
+def link_keyfile(keyfile_path: Path) -> None:
+    if valid_keyfile_exists():
+        xdg_keyfile = get_keyfile_path()
+        cli_logger.info(
+            f"Keyfile detected at {xdg_keyfile}. Please use `ethpm auth unlink` to delete this "
+            "keyfile before linking a new one."
+        )
+    else:
+        import_keyfile(keyfile_path)
+        address = get_authorized_address()
+        cli_logger.info(
+            f"Keyfile stored for address: {address}\n"
+            "It's now available for use when its password is passed in with the "
+            "`--keyfile-password` flag."
+        )
+
+
+def unlink_keyfile() -> None:
+    if not valid_keyfile_exists():
+        cli_logger.info("Unable to unlink keyfile: empty keyfile found.")
+    else:
+        keyfile_path = get_keyfile_path()
+        address = get_authorized_address()
+        keyfile_path.write_text("")
+        cli_logger.info(f"Keyfile removed for address: {address}")
+
+
+def init_keyfile() -> None:
+    if valid_keyfile_exists():
+        cli_logger.info(
+            f"Keyfile detected. Please use `ethpm auth unlink` to delete this "
+            "keyfile before initializing a new one."
+        )
+        return
+
+    cli_logger.info(PRIVATE_KEY_WARNING)
+    agreement = parse_bool_flag(
+        "Are you sure you want to proceed with initializing a keyfile? "
+    )
+    if not agreement:
+        cli_logger.info("Aborting keyfile initialization.")
+        return
+
+    private_key = to_bytes(text=input("Please enter your 32-length private key: "))
```

Comment:
I don't think this is ok.  I know you are probably expecting people to copy/paste which *might* be fine if it could be enforced, but this will also result in people typing the key in, or pasting it with an accidental newline at the end or typing it in its hexidecimal representation, or any number of insane things we can't think of....  I think that if someone wants to bring in a raw private key we should require them to do the heavy lifting of creating the keyfile themselves so that we're not responsible for mistakes by encouraging this pattern... thoughts?

cc @holiman